test/components/enketo-iframe.spec.js

```diff
@@ -0,0 +1,59 @@
+import sinon from 'sinon';
+import EnketoIframe from '../../src/components/enketo-iframe.vue';
+import NotFound from '../../src/components/not-found.vue';
+import { mount } from '../util/lifecycle';
+import { mockRouter } from '../util/router';
+import { wait } from '../util/util';
+
+const enketoId = 'sCTIfjC5LrUto4yVXRYJkNKzP7e53vo';
+
+const mountComponent = (props) => mount(EnketoIframe, {
+  props,
+  container: {
+    router: mockRouter()
+  },
+  attachTo: document.body
+});
+
+describe('EnketoIframe', () => {
+  [
+    { desc: 'all null', enketoId: null, actionType: null, instanceId: null },
+    { desc: 'invalid actionType', enketoId, actionType: 'delete', instanceId: null },
+    { desc: 'null instanceId with edit', enketoId, actionType: 'edit', instanceId: null },
+  ].forEach(t => {
+    it(`renders NotFound component when props are invalid - ${t.desc}`, () => {
+      const { desc, ...props } = t;
+      const wrapper = mountComponent(props);
+      wrapper.findComponent(NotFound).exists().should.be.true;
+    });
+  });
+
+  it('should update title to Page Not Found when props are invalid', () => {
+    mountComponent();
+    document.title.should.match(/Page Not Found/);
+  });
+
+  it('renders iframe with correct src when props are valid', () => {
+    const wrapper = mountComponent({ enketoId, actionType: 'edit', instanceId: 'test-instance' });
+    const iframe = wrapper.find('iframe');
+    iframe.exists().should.be.true;
+    iframe.attributes('src').should.contain(`/enketo-passthrough/edit/${enketoId}?instance_id=test-instance`);
+  });
+
+  it('redirects on submissionsuccess message with return_url', async () => {
+    const wrapper = mountComponent({ enketoId, actionType: 'new', instanceId: 'test-instance' });
+    const iframe = wrapper.find('iframe');
+
+    wrapper.vm.$route.query = { return_url: 'http://localhost/projects/1' };
+
+    wrapper.vm.$router.push = sinon.fake();
+
+    iframe.element.contentWindow.eval(`
+      window.parent.postMessage(JSON.stringify({ enketoEvent: 'submissionsuccess' }), "*");
```

Comment:
I always feel a little nervous when I see `eval()`, even though I know it's safe here. 😅 Is there another way we could call `postMessage()` though? Is `eval()` needed?

Reply:
Is it possible to do something like:

```js
iframe.element.contentWindow.parent.postMessage(/*...*/);
```

Or is that restricted in some way?

Reply:
I wanted to send message from inside the iframe, but I guess this is also fine.

Reply:
I guess injecting a `<script>` tag doesn't feel too different from `eval()` to me. But maybe at the end of the day, both are actually fine. If what you're doing feels like the right approach to you because it's inside the iframe, I'm happy for us to run with that.